libtransmission/tr-lpd.cc

```diff
@@ -508,28 +627,40 @@ class tr_lpd_impl final : public tr_lpd
      * matter). A listening client on the same network might react by adding us to his
      * peer pool for torrent t.
      */
-    bool sendAnnounce(std::vector<std::string_view> const& info_hash_strings)
+    bool sendAnnounce(tr_address_type ip_protocol, std::vector<std::string_view> const& info_hash_strings)
     {
-        auto const announce = makeAnnounceMsg(cookie_, mediator_.port(), info_hash_strings);
+        TR_ASSERT(tr_address::is_valid(ip_protocol));
+        if (!tr_address::is_valid(ip_protocol))
+        {
+            return false;
+        }
+
+        if (mcast_sockets_[ip_protocol] == TR_BAD_SOCKET)
+        {
+            return true;
+        }
+
+        auto const announce = makeAnnounceMsg(ip_protocol, cookie_, mediator_.port(), info_hash_strings);
         TR_ASSERT(std::size(announce) <= MaxDatagramLength);
         auto const res = sendto(
-            mcast_socket_,
+            mcast_sockets_[ip_protocol],
             std::data(announce),
             std::size(announce),
             0,
-            reinterpret_cast<sockaddr const*>(&mcast_addr_),
-            sizeof(mcast_addr_));
-        auto const sent = res == static_cast<int>(std::size(announce));
-        return sent;
+            ip_protocol == TR_AF_INET ? reinterpret_cast<sockaddr const*>(&mcast_addr_) :
+                                        reinterpret_cast<sockaddr const*>(&mcast6_addr_),
+            ip_protocol == TR_AF_INET ? sizeof(mcast_addr_) : sizeof(mcast6_addr_));
+        return res == static_cast<int>(std::size(announce));
     }
 
     std::string const cookie_ = makeCookie();
     Mediator& mediator_;
-    tr_socket_t mcast_socket_ = TR_BAD_SOCKET; /**multicast socket */
-    libtransmission::evhelpers::event_unique_ptr event_;
+    std::array<tr_socket_t, NUM_TR_AF_INET_TYPES> mcast_sockets_ = { TR_BAD_SOCKET, TR_BAD_SOCKET }; /**multicast sockets */
+    std::array<libtransmission::evhelpers::event_unique_ptr, NUM_TR_AF_INET_TYPES> events_;
 
     static auto constexpr MaxDatagramLength = size_t{ 1400 };
     sockaddr_in mcast_addr_ = {}; /**<initialized from the above constants in init() */
+    sockaddr_in6 mcast6_addr_ = {}; /**<initialized from the above constants in init() */
```

Comment:
nit: let's avoid trailing `/**` comments which aren't well supported? Could be changed to trailing `//` comments, or trailing `///<` documentation comments, or previous-line `///` documentation comments.

Reply:
a066bc6

Reply:
Still one on `mcast_sockets_`, but it's not a blocker for this PR.

Reply:
53e4c80